Improve wifi coex doc (#952)
nrfconnect/sdk-nrfxlib: mpsl/doc/cx.rst

.. _mpsl_cx:

Short-Range Protocols External Radio Coexistence
################################################

.. contents::
   :local:
   :depth: 2

The radio coexistence feature allows the application to interface with several types of packet traffic arbiters (PTAs).

> PTAs arbitrate the requested radio operations between all radios to avoid interference, providing better radio performance to devices using multiple interfering radios simultaneously, like a combination of IEEE 802.15.4, Bluetooth LE, and Wi-Fi.

Reviewer: ```suggestion
PTAs arbitrate the requested radio operations between all radios to avoid interference, providing better radio performance to devices using multiple interfering radios simultaneously, like a combination of IEEE 802.15.4, Bluetooth® Low Energy (LE), and Wi-Fi.
```

Author: fixed

.. _mpsl_cx:

Short-Range Protocols External Radio Coexistence
################################################

.. contents::
   :local:
   :depth: 2

> The radio coexistence feature allows the application to interface with several types of packet traffic arbiters (PTAs).

Author: Cant see difference

> This allows for better radio performance for devices using multiple interfering radios simultaneously, like a combination of IEEE 802.15.4 Bluetooth LE and Wi-Fi.

Reviewer 2: Missing comma ;)

Author: fixed

> The PTA arbitrates the requested radio operations between all radios to avoid interference, increasing the performance of the radio protocols enabled simultaneously.

Reviewer 3: ```suggestion
PTAs arbitrate the requested radio operations between all radios to avoid interference, providing better radio performance to devices using multiple interfering radios simultaneously, like a combination of IEEE 802.15.4, Bluetooth LE, and Wi-Fi.
```

> The exact arbitration algorithm is dependent on the specific PTA used.

Reviewer 3: ```suggestion
The arbitration algorithm used can vary between PTAs.
```

.. note::
  The radio coexistence feature is not needed for the arbitration between |BLE| and IEEE 802.15.4 in dynamic multiprotocol applications, as the dynamic multiprotocol does not use PTA for the arbitration.

> Principle of operation

Reviewer 3: ```suggestion
Overview
********
```

> **********************

Reviewer 3: ```suggestion
```

The radio coexistence feature allows short-range protocol drivers (e.g. IEEE 802.15.4, Bluetooth LE) to communicate with the packet traffic arbiter (PTA) using :ref:`MPSL CX API <mpsl_api_sr_cx>`.

> The MPSL CX API is hardware-agnostic and separates implementation of a protocol diver from an implementation specific for given PTA.

Reviewer 2: diver --> driver

Author: fixed

> The MPSL CX API is hardware-agnostic and separates implementation of a protocol driver from an implementation specific for given PTA.

Reviewer 3: ```suggestion
The MPSL CX API is hardware-agnostic and separates the implementation of the protocol driver from an implementation specific to given PTA.
```

> To perform any radio operation, the protocol drivers must first request appropriate access to the medium from the PTA.

Reviewer 3: ```suggestion
To perform any radio operation, the protocol drivers must first request the appropriate access to the medium from the PTA.
```

The request informs the PTA implementation about which radio operations it wants to perform at that moment or shortly after, and what is the priority of the operation.
The radio operation can be executed only when the PTA grants it.
When the PTA revokes access to the medium during an ongoing radio operation, the radio protocol must abort this operation immediately.

> When the radio operation is finished the protocol driver must release the requested operation.

Reviewer 3: ```suggestion
When the radio operation is finished, the protocol driver must release the requested operation.
```

The implementation of MPSL CX translates the request invoked by radio protocol to hardware signals compatible with the PTA available in the system.
It also translates the hardware signals received from the PTA indicating which radio operations are allowed at that moment, and it passes this information to the radio protocol.

.. note::
  To learn more about the arbitration process during various transmission and reception modes for the IEEE 802.15.4 Radio Driver, see :ref:`Wi-Fi Coexistence module (PTA support) <rd_coex>`

Selecting CX Implementation
***************************

The :ref:`mpsl` itself does not provide any implementation of the CX interface.

> For implementations present in |NCS| please refer to :ref:`ug_radio_coex`.

Reviewer 3: ```suggestion
For details on the implementations present in the |NCS|, see :ref:`ug_radio_coex`.
```

> If an application needs to use CX it must call :c:func:`mpsl_cx_interface_set()` during the system initialization.

Reviewer 3: ```suggestion
An application that needs to use CX must call :c:func:`mpsl_cx_interface_set()` during the system initialization.
```

> Initialization of any resources needed by selected CX implementation is out of scope of :ref:`mpsl` and should also be done during the system initialization.

Reviewer 3: ```suggestion
The initialization of any resource needed by the selected CX implementation is not in scope of :ref:`mpsl` and must also be done during the system initialization.
```

.. note::

>   In the |NCS| the selection of a CX implementation and appropriate initialization is done automatically by use of Kconfig and Device Tree settings.

Reviewer 3: ```suggestion
  In the |NCS|, the selection of a CX implementation and its appropriate initialization is done automatically, using Kconfig and Device Tree configuration options.
```

  Please refer to :ref:`ug_radio_coex`.

Using CX API in a protocol driver
*********************************

> To use the CX API in a protocol driver you must follow :ref:`MPSL CX API <mpsl_api_sr_cx>` defined by :file:`nrfxlib/mpsl/include/protocol/mpsl_cx_protocol_api.h`.

Reviewer 3: ```suggestion
To use the CX API in a protocol driver, you must follow the :ref:`MPSL CX API <mpsl_api_sr_cx>` defined by :file:`nrfxlib/mpsl/include/protocol/mpsl_cx_protocol_api.h`.
```

Implementing CX API
*******************

> For implementations of the MPSL CX API for certain PTAs already supported by |NCS| please refer to :ref:`ug_radio_coex_mpsl_cx_based`.

Reviewer 3: ```suggestion
For details on the implementations of the MPSL CX API for certain PTAs already supported by the |NCS|, see :ref:`ug_radio_coex_mpsl_cx_based`.
```

> If your PTA is not supported yet, please refer to :ref:`ug_radio_mpsl_cx_custom` for guidelines on how to add your own implementation.

Reviewer 3: ```suggestion
If your PTA is not supported yet, see :ref:`ug_radio_mpsl_cx_custom` for guidelines on how to add your own implementation.
```
